test_fixtures/android_news_feed.html

```diff
@@ -0,0 +1,42 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <script src="../scripts/gentest/test_helper.js"></script>
+  <link rel="stylesheet" type="text/css" href="../scripts/gentest/test_base_style.css">
+  <title>
+    Test description
+  </title>
+<head/>
+<body>
+
+  <div id="test-root" layout="width: 1080; height: 444; top: 0; left: 0;" style="align-content: stretch; flex-shrink: 0; width: 1080px; " >
+    <div layout="width: 1080; height: 444; top: 0; left: 0;" style="flex-shrink: 0; " >
+      <div layout="width: 1080; height: 444; top: 0; left: 0;" style="align-content: stretch; " >
+        <div layout="width: 1080; height: 226; top: 0; left: 0;" style="align-content: stretch; " >
+          <div layout="width: 1044; height: 202; top: 24; left: 36;" style="flex-direction: row; align-items: flex-start; align-content: stretch; margin-top: 24px; margin-start: 36px; " >
```

Comment:
The "margin-start" here is not a valid CSS property, and Yoga's test generator did some funky things with it. But it's effectively the same as "margin-inline-start" for any case that isn't a vertical writing direction (which I don't think Yoga or Taffy support).

Reply:
Ah, that's useful to know. Taffy actually doesn't support `margin-inline-start` either (it has `margin-left`). I'm guessing Yoga also doesn't support this and the test generator just converts it to left/right?

Reply:
I did a test run overwriting all of Taffy's test with Yoga's (where there as an equivalent) to see what the diff would be like, and there were quite a few differences. In addition to the start/end vs. left/right thing, the main one's were due to Yoga defaulting to `flex-direciton: column` while Taffy defaults to `flex-direction: row`. Many of a Taffy's tests had an explicit flex-direction added to account for this. And some of them had width/height swapped to account for it in a different way. I suspect we'll need to add explicit `flex-direction` everywhere if we want to unify the test suites.

Reply:
> Ah, that's useful to know. Taffy actually doesn't support `margin-inline-start` either (it has `margin-left`). I'm guessing Yoga also doesn't support this and the test generator just converts it to left/right?

Yoga does have support for flow-relative margin, padding, etc. The API at the Yoga level looks like `YGNodeStyleSetMargin`, accepting a `YGEdge`, which can be (in order of precedence) a cardinal direction (e.g. `YGEdgeLeft`), a flow-relative direction (e.g. `YGEdgeStart`), an axis (e.g. `YGEdgeHorizontal`), or `YGEdgeAll`.

React Native historically historically mapped `YGEdgeStart` to `marginStart`, etc, but we recently added `marginInlineStart`, `marginBlockStart`, etc to be conformant with browsers.

Reply:
Re defaults, Yoga supports a `YGConfig` option `UseWebDefaults`. There are a couple other differences in the default style which that overrides. My historical understanding was that it was to eventually become the default, but the work never materialized.

Reply:
> a flow-relative direction (e.g. YGEdgeStart), an axis (e.g. YGEdgeHorizontal)"

Wouldn't you need a flow-relative direciton AND an axis to fully specify an edge? (`YGEdgeStart` by itself is ambiguous. You'd need `YGEdgeInlineStart`, no?).

> The API at the Yoga level looks like YGNodeStyleSetMargin, accepting a YGEdge, which can be (in order of precedence) a cardinal direction (e.g. YGEdgeLeft), a flow-relative direction (e.g. YGEdgeStart), an axis (e.g. YGEdgeHorizontal), or YGEdgeAll.

Do you actually store the flow-relative styles (e.g. `margin-inline-start`) separately from the absolute styles (e.g. `margin-left`)? i.e. would it compute the correct layout if a user starts toggling the RTL flag after setting the styles? And if it does do that, how are you dealing with the effective doubling of memory usage that this entails? Are you just eating that cost? I tried looking through the yoga source, and it looked to me that the size of the margin property in`YGStyle` was equal to the number of variants in the `YGEdge`, but given the variants in YGEdge that didn't make any sense to me.

The templated C++ type definitions were going over my head a bit though, so it's possible I wasn't understanding this correctly.

Reply:
> Re defaults, Yoga supports a YGConfig option UseWebDefaults. There are a couple other differences in the default style which that overrides. My historical understanding was that it was to eventually become the default, but the work never materialized.

Perhaps it would make sense to make this the default specifically for the test suite? (updating the test fixtures accordingly so that the actual tests don't change).

Reply:
> Wouldn't you need a flow-relative direciton AND an axis to fully specify an edge? (YGEdgeStart by itself is ambiguous. You'd need YGEdgeInlineStart, no?).

Start/End only ever refer to the horizontal axis, and just picks Left/Right based on RTL. But yeah,  the style has distinct inputs for padding, margin, etc per YGEdge, and a specific precedence when multiple  inputs are present which may overlap the same physical edge. It would be reevaluated if laid out as RTL. Yoga also weirdly treats both YGEdgeLeft and YGEdgeStart as effecting the right side of a box with flexDirection: "row-reverse" which is inconsistent with browsers. 

Yoga's precedence logic based on specificity is one difference from browsers, where CSS properties are ordered and if there is overlap, the last assignment wins (e.g. "margin" takes precedence over "marginLeft" if it comes after).

Reply:
> Perhaps it would make sense to make this the default specifically for the test suite? (updating the test fixtures accordingly so that the actual tests don't change).

I think updating that would probably make sense. The existing passing fixtures must be explicit enough to set styles that are interpreted the came between both browsers and Yoga in its normal layout mode, but I don't know if we want that requirement.

Reply:
> If you're having to add column direction, was Yoga's test generator doing that in the outer HTML too?

I haven't actually checked that yet. I've gone through and added all the `flex-direction: column`s. But actually, all of those tests were already passing. So it's possible that the missing flex-directions weren't actually making a meaningful difference to the tests. I'd still be more comfortable if the tests are unambiguously specified though.

Reply:
> Start/End only ever refer to the horizontal axis, and just picks Left/Right based on RTL.

Ah, I guess that makes sense given that IIRC Yoga supports RTL but not vertical writing modes. The variants are confusingly named though! I'm guessing that might the sort of thing it might be more possible to make breaking changes to (as it will turn into a compile error)?

> Yoga also weirdly treats both YGEdgeLeft and YGEdgeStart as effecting the right side of a box with flexDirection: "row-reverse" which is inconsistent with browsers.

That is weird. As documented in https://github.com/DioxusLabs/taffy/issues/307, it does work that way for the `flex-start` alignment value, but there is also a separate `start` which isn't affected by the the `flex-direction`, and it definitely shouldn't do that for `left`!

Reply:
> If you're having to add column direction, was Yoga's test generator doing that in the outer HTML too?

@NickGerleman I looked into this, and the stylesheet that Yoga uses when generating tests applies this default. See https://github.com/facebook/yoga/blob/main/gentest/test-template.html#L28